diff --git a/perf/cairo-analyse-trace.c b/perf/cairo-analyse-trace.c
index 9941486..9e70037 100644
--- a/perf/cairo-analyse-trace.c
+++ b/perf/cairo-analyse-trace.c
@@ -76,7 +76,7 @@ basename_no_ext (char *path)
 {
     char *dot, *name;
 
-    name = basename (path);
+    name = strdup("no_basename_syscall");//basename (path);
 
     dot = strchr (name, '.');
     if (dot)
diff --git a/perf/cairo-perf-trace.c b/perf/cairo-perf-trace.c
index f27f8e4..76cc217 100644
--- a/perf/cairo-perf-trace.c
+++ b/perf/cairo-perf-trace.c
@@ -77,7 +77,7 @@ basename_no_ext (char *path)
 {
     char *dot, *name;
 
-    name = basename (path);
+    name = strdup("no_basename_syscall");//basename (path);
 
     dot = strrchr (name, '.');
     if (dot)
